packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2072,8 +2071,22 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
   TextSelection _getWordAtOffset(TextPosition position) {
     debugAssertLayoutUpToDate();
     final TextRange word = _textPainter.getWordBoundary(position);
+    final int effectiveOffset;
+    switch (position.affinity) {
+      case TextAffinity.upstream:
+        // upstream affinity is effectively -1 in text position.
+        effectiveOffset = position.offset - 1;
+        break;
+      case TextAffinity.downstream:
+        effectiveOffset = position.offset;
+        break;
+    }
     // When long-pressing past the end of the text, we want a collapsed cursor.
-    if (position.offset >= word.end) {
+    // This also needs to account for cases where user tap at the end of text.
```

Comment:
directionality should be fine as the skia should handle that for us? The problem I facing here is I can't differentiate whether user taps the upstream of the end of the text or far away to the right from the end of the text. I think the correct behavior is we should still select word for the former and collapse at the end for the latter.

Since we can't tell it with the current API, I made it always collapse at the end, since that is what the original behavior is